Refactor `Layout`, `HtmlAttributes`, `HtmlList` and `BasePath` helpers

gsteel opened the pull request:
Also adds a factory for the plugin manager.

Generally speaking, the helpers have…

- Deprecations removed
- Inheritance removed
- Types added
- Final keyword added

gsteel committed:
Refactor `ViewModel` Helper

General cleanup, adds types, removes inheritance and implement `StatefulHelperInterface`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add `composer validate --strict` to QA checks

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `Layout` Helper

- Removes inheritance hierarchy
- Removes `setTemplate()` method because it is infeasible to retrieve an instance from a view context without setting the template anyway.
- Changes constructor to require a composed ViewModel helper
- Adds a factory for the helper

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds a factory for the helper plugin manager.

This codifies the convention of configuring view helpers under the `view_helpers` config key

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `HtmlAttributes`

Removes inheritance hierarchy and adds the `final` keyword

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `HtmlList` Helper

Removes inheritance hierarchy and adds the `final` keyword

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Use the `EscapeHelperFactory` to create `HtmlList` and `GravatarImage` helpers so that the escaper in the container is reused

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Re-order helpers alphabetically and remove the v2 normalized aliases

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `BasePath` helper

- Removes inheritance
- Adds final keyword
- Adds types
- Implements `StatefulHelperInterface`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refresh lock and baseline

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update migration guide

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -15,11 +15,29 @@ final class ConfigProvider
+            /**
+             * The top-level configuration key for defining custom view helpers.
+             *
+             * This option should use the `ServiceManagerConfiguration` array format
```
Should we type against it then?

gsteel replied:
Yes! I also sealed the shape 👍

froschdesign commented on src/ConfigProvider.php:
```diff
@@ -15,11 +15,29 @@ final class ConfigProvider
+             */
+            'view_helpers'       => [],
             'view_manager'       => [
+                /**
+                 * Encoding passed to the Escaper and possibly used in other view-related configuration
+                 */
```
laminas-mvc does not set the encoding via the view manager. Can we move this to `view_helper_config`?

froschdesign commented on src/ConfigProvider.php:
```diff
@@ -15,11 +15,29 @@ final class ConfigProvider
+                 */
                 'encoding' => 'utf-8',
+
+                /**
+                 * The base path is provided to the BasePath view helper. This is the historic location for this
+                 * configuration item defined in MVC apps.
+                 */
+                'base_path' => null,
```
laminas-mvc cannot use the new major version of laminas-view, so the question arises if we need the configuration key `view_manager` at all.
Related to https://github.com/laminas/technical-steering-committee/pull/223

gsteel replied:
I've moved some stuff to `view_helper_config` anyway, and retained searching in `view_manager` for fewer BC issues.

#281 refactors template resolvers, so we'll be writing factories soon to set those up. MVC puts all of this under `view_manager`, Mezzio under `templates` (Both at the top-level).

I'm starting to wonder if we could do with a top-level key that is specific to this library such as `laminas-view` where we nest everything else…

gsteel committed:
Move all view configuration to `view_helper_config`

Preserves and tests for finding config options in "legacy" locations for `encoding` and `base_path`

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/Helper/HtmlAttributes.php:
```diff
@@ -9,18 +9,14 @@
-    private Escaper $escaper;
+    private readonly Escaper $escaper;
 
     public function __construct(?Escaper $escaper = null)
```
BTW, since these helpers are always built via DIC, perhaps make the ctor argument mandatory, and remove instantiation?

It's just an idea / something I noticed

gsteel replied:
Good point. If Escaper's constructor ever changes, however unlikely, this ensures deps are fully delegated 👍

Ocramius commented on src/Helper/Service/BasePathFactory.php:
```diff
@@ -4,32 +4,27 @@
 
 final class BasePathFactory
 {
+    /** @psalm-suppress MixedAssignment */
```
Eh, let's suppress per-statement, not just everywhere? Mixed assignments are very sneaky 

Fine to dismiss this review comment, if you believe it's not as important.

gsteel replied:
You're right of course! I've imported the "View Config" array shape and forced that var on `$container->get('config')`

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -15,14 +15,29 @@ final class ConfigProvider
-                'encoding' => 'utf-8',
-            ],
+            'dependencies' => $this->getDependencies(),
+            /**
```
Since we are introducing them now, can we document the type of additional keys in this array at type level?

gsteel committed:
Make escaper a mandatory constructor property.

Relying completely on the DI container for the escaper instance shields us from the potential for change in the external library

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add more precise types for the configuration array shape.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Make the escape mandatory in the `HtmlAttributes` constructor.

The dedicated factory for this helper is now unnecessary and is removed in favour of `EscapeHelperFactory`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Simplifications from better types in `ConfigProvider`

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius reviewed (approved):
:shipit:
